src/otr/otr-ops.c

```diff
@@ -0,0 +1,362 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,USA
+ */
+
+#include "common.h"
+
+#include "signals.h"
+#include "levels.h"
+#include "printtext.h"
+#include "fe-windows.h"
+
+#include "key.h"
+#include "module.h"
+#include "otr-formats.h"
+#include "irssi-otr.h"
+
+static OtrlPolicy OTR_DEFAULT_POLICY = OTRL_POLICY_MANUAL | OTRL_POLICY_WHITESPACE_START_AKE;
+
+/*
+ * Return default policy for now.
+ */
+static OtrlPolicy ops_policy(void *opdata, ConnContext *context)
+{
+	return OTR_DEFAULT_POLICY;
+}
+
+/*
+ * Request for key generation.
+ *
+ * The lib actually expects us to be finished before the call returns. Since
+ * this can take more than an hour on some systems there isn't even a point in
+ * trying...
+ */
+static void ops_create_privkey(void *opdata, const char *accountname,
+		const char *protocol)
+{
+	key_gen_run(user_state_global, accountname);
+}
+
+/*
+ * Inject OTR message.
+ */
+static void ops_inject_msg(void *opdata, const char *accountname,
+		const char *protocol, const char *recipient, const char *message)
+{
+	SERVER_REC *server = opdata;
+
+	IRSSI_OTR_DEBUG("Inject msg:\n[%s]", message);
+	irssi_send_message(server, recipient, message);
+}
+
+/*
+ * Gone secure.
+ */
+static void ops_secure(void *opdata, ConnContext *context)
+{
+	char ownfp[OTRL_PRIVKEY_FPRINT_HUMAN_LEN];
+	char peerfp[OTRL_PRIVKEY_FPRINT_HUMAN_LEN];
+	SERVER_REC *server = opdata;
+	struct otr_peer_context *opc;
+
+	g_assert(context != NULL);
+	/* This should *really* not happened */
+	g_assert(context->msgstate == OTRL_MSGSTATE_ENCRYPTED);
+
+	printformat(server, context->username, MSGLEVEL_CLIENTCRAP, TXT_OTR_SESSION_SECURE);
+	otr_status_change(server, context->username, OTR_STATUS_GONE_SECURE);
+
+	opc = context->app_data;
+	opc->active_fingerprint = context->active_fingerprint;
+
+	if (otrl_context_is_fingerprint_trusted(context->active_fingerprint)) {
+		/* Secure and trusted */
+		return;
+	}
+
+	/* Not authenticated. Let's print out the fingerprints for comparison. */
+	otrl_privkey_hash_to_human(peerfp, context->active_fingerprint->fingerprint);
+	otrl_privkey_fingerprint(user_state_global->otr_state, ownfp, context->accountname, OTR_PROTOCOL_ID);
+
+	printformat(server, context->username, MSGLEVEL_CLIENTCRAP, TXT_OTR_SESSION_UNAUTHENTICATED_WARNING);
+	printformat(server, context->username, MSGLEVEL_CLIENTCRAP, TXT_OTR_FP_INFO, server->nick, ownfp);
+	printformat(server, context->username, MSGLEVEL_CLIENTCRAP, TXT_OTR_FP_INFO, context->username, peerfp);
+}
+
+/*
+ * Gone insecure.
+ */
+static void ops_insecure(void *opdata, ConnContext *context)
```

Comment:
This isn't something you can address but i feel the need to complain that libotr doesn't emit this op ever, and my patch never got merged https://bugs.otr.im/lib/libotr/issues/48

Reply:
:cry: